### PR TITLE
First modifications for 2018 

### PR DESCRIPTION
**Pull request by dzuolo**
Added branches in bigTree, skimNtuple.cpp and smallTree to store the new deepTauID scores for dau1 and dau2 and the deepFlavor scores for all the used jets.
Created a new skimNtuple config file for 2018. At the moment is a copy of the 2017 one with a new flag ("useDeepFlavor = false") to use deepFlavor score to order bjets.
Added deepFlavor SF csv file for 2018.
Added deepCSV and deepFlavor WPSet to bTagSF.cc
See https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuidePFTauID#Running_of_the_DeepTauIDs_ver_20 for the new DNN based TauID.
See https://twiki.cern.ch/twiki/bin/viewauth/CMS/BtagRecommendation102X for bTagging related changes.

### CODE REVIEW

**Comment by dzuolo** [comment]
Added first IdIso and Trigger SF files from https://github.com/CMS-HTT/LeptonEfficiencies (Electron and Muon Trigger And IdIso SF files) and https://github.com/cms-tau-pog/TauTriggerSFs (Tau Trigger SF Files). Tau ID SF are taken from https://twiki.cern.ch/twiki/bin/viewauth/CMS/TauIDRecommendation13TeV#Tau_ID_efficiency

**Comment by dzuolo** [comment]
Added txt file, data PU distribution and instructions in the code for 2018 data sample PUReweight.
Fixed some bugs discovered after trying to submit skim jobs.

**Comment by dzuolo** [comment]
DeepTau branches temporary commented: check if the output is floating or integer and eventually fix in LLR

File: test/skimNtuple.cpp [hunk 1]
```diff
@@ -967,7 +968,9 @@ int main (int argc, char** argv)
   string bTag_effFile = gConfigParser->readStringOption("bTagScaleFactors::effFile") ;
   cout << "B Tag SF file: " << bTag_SFFile << endl;
   //bTagSF bTagSFHelper (bTag_SFFile, bTag_effFile, "", "80X_MORIOND_2017"); // third field unused, but could be needed to select efficiencies for different selection levels
-  bTagSF bTagSFHelper (bTag_SFFile, bTag_effFile, "", "94X_DeepCSV_V1"); // third field unused, but could be needed to select efficiencies for different selection levels
+  bTagSF bTagSFHelper (bTag_SFFile, bTag_effFile, "", "102X_DeepCSV_V1"); // third field unused, but could be needed to select efficiencies for different selection levels
+  if(useDeepFlavor)
+  	bTagSFHelper.SetWPset("102X_DeepFlavor_V1"); // third field unused, but could be needed to select efficiencies for different selection levels
```
Comment (francescobrivio) [inline test/skimNtuple.cpp:990]: the "bTag_SFFile" and "bTag_effFile" are read directly from the cfg right?
maybe it's bettere to have different files for deepCSV and deepFlavor also in the config, so we do not have to change them every time, you can just change the "useDeepFlavor" flag and the rest is all automatic

File: test/skimNtuple.cpp [hunk 2]
```diff
@@ -3373,7 +3497,7 @@ int main (int argc, char** argv)
 
         bool bPairFound = false;
         int njets = jets_and_sortPar.size();
-        if (jets_and_sortPar.at(njets-2).first>0.4941) bPairFound = true; // medium WP is: 0.8484 for 2016 CSV, 0.4941 for 2017 DeepCSV
+        if (jets_and_sortPar.at(njets-2).first>0.4184) bPairFound = true; // medium WP is: 0.8484 for 2016 CSV, 0.4941 for 2017 DeepCSV, 0.4184 for 2018 DeepCSV,  0.2770 for 2018 DeepFlavor
```
Comment (francescobrivio) [inline test/skimNtuple.cpp]: shouldn't this be like the other deepCSV/deepFlavor choice? 
something like:
```
if(useDeepFlavor)
   if (jets_and_sortPar.at(njets-2).first>0.2770) bPairFound = true;  
else
   if (jets_and_sortPar.at(njets-2).first>0.4184) bPairFound = true;
```

File: config/skim_2018_sync_mib.cfg [hunk 3]
```diff
@@ -0,0 +1,241 @@
+### how to run on my file:
+### skimNtuple.exe filelist_sync13Sett.txt output_HH_sync_14Set.root 10.0 0  config/skim_2016_sync.cfg  0  1.0 -999.0 0  0  0 0 1
```
Comment (francescobrivio) [inline config/skim_2018_sync_mib.cfg]: lines 1 and 2 can be deleted I think...

File: config/skim_2018_sync_mib.cfg [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+
+[bRegression]
+computeBregr = false
+weights = /gwpool/users/brivio/Hhh_1718/syncFeb2018/CMSSW_7_4_7/src/KLUBAnalysis/weights/TMVARegression_BDTG.weights.xml
+#weights = /home/llr/cms/amendola/CMSSW_7_4_7/src/KLUBAnalysis/weights/TMVARegression_BDTG.weights.xml
+
+[bTagScaleFactors]
+#SFFile  = /gwpool/users/dzuolo/HbbtautauAnalysis2018/CMSSW_9_0_0/src/KLUBAnalysis/weights/DeepJet_102XSF_V1.csv
+SFFile  = /gwpool/users/dzuolo/HbbtautauAnalysis2018/CMSSW_9_0_0/src/KLUBAnalysis/weights/DeepCSV_102XSF_V1.csv
```
Comment (francescobrivio) [inline config/skim_2018_sync_mib.cfg]: we should create 2 different SFFiles and in skimNtuple read one or the other based on the "useDeepFlavor" flag:
```
SFFile_deepCSV
SFFfile_deepFlavor
```

File: config/skim_2018_sync_mib.cfg [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+#effFile  = /gwpool/users/dzuolo/HbbtautauAnalysis2018/CMSSW_9_0_0/src/KLUBAnalysis/weights/DeepJet_102XEff_V1.root
+#effFile  = /gwpool/users/dzuolo/HbbtautauAnalysis2018/CMSSW_9_0_0/src/KLUBAnalysis/weights/DeepCSV_102XEff_V1.root
+effFile  = /gwpool/users/dzuolo/HbbtautauAnalysis2018/CMSSW_9_0_0/src/KLUBAnalysis/weights/bTagEfficiencies_94X_2018WPs_Fall17MC_cleanJets.root
```
Comment (francescobrivio) [inline config/skim_2018_sync_mib.cfg]: same comment as line 187

File: interface/PUReweight.h [hunk 6]
```diff
@@ -24,6 +24,10 @@ class PUReweight {
   Type theType;
   TString theFilename;
 
+  // 2018 Samples
+  std::vector<float> hT2018_MCAutumun18_Data59c7fb; // Full Run2018
```
Comment (francescobrivio) [inline interface/PUReweight.h]: Autumun --> Autumn

File: interface/PUReweight.h [hunk 7]
```diff
@@ -24,6 +24,10 @@ class PUReweight {
...
+  std::vector<float> hT2018_MCAutumun18_Data59c7fb_special;  // Full Run2018 - weights read from external txt file
```
Comment (francescobrivio) [inline interface/PUReweight.h]: Autumun --> Autumn

File: interface/smallTree.h [hunk 8]
```diff
@@ -632,6 +654,13 @@ struct smallTree
       m_smallT->Branch ("dau1_CUTiso", &m_dau1_CUTiso, "dau1_CUTiso/I") ;
       m_smallT->Branch ("dau1_antiele", &m_dau1_antiele, "dau1_antiele/I") ;
       m_smallT->Branch ("dau1_antimu", &m_dau1_antimu, "dau1_antimu/I") ;
+      //m_smallT->Branch ("dau1_deepTauVsJet",m_dau1_deepTauVsJet,"dau1_deepTauVsJet/I");
+      //m_smallT->Branch ("dau1_deepTauVsEle",m_dau1_deepTauVsEle,"dau1_deepTauVsEle/I");
+      //m_smallT->Branch ("dau1_deepTauVsMu",m_dau1_deepTauVsMu,"dau1_deepTauVsMu/I");
+     
+     
+     
+
```
Comment (francescobrivio) [inline interface/smallTree.h:660]: delete a couple of these empty lines

File: src/PUReweight.cc [hunk 9]
```diff
@@ -64,6 +64,8 @@ float PUReweight::weight(int MC, int target, int input, TString filename) {
       return hT2017_MCFall17_Data13c4fb.at(input) ;
     } else if (MC==2017 && target==413) {
       return hT2017_MCFall17_Data41c3fb.at(input) ;         // Full 2017
+    } else if (MC==2018 && target==597) {
+      return hT2018_MCAutumun18_Data59c7fb.at(input) ;         // Full 2018
```
Comment (francescobrivio) [inline src/PUReweight.cc]: Autumun --> Autumn

File: src/PUReweight.cc [hunk 10]
```diff
@@ -74,6 +76,16 @@ float PUReweight::weight(int MC, int target, int input, TString filename) {
...
       }
+    } else if (MC==2018 && target==999) {                   // Ful 2018 with weights for each sample read from external txt file
+      if (filename == TString(""))
+      {
+        std::cout << "ERROR: PUReweight - You requested special weights but didn't pass the external file: aborting !!! " << std::endl;
+        abort();
+      }
+      else
+      {
+        return hT2018_MCAutumun18_Data59c7fb_special.at(input) ;
```
Comment (francescobrivio) [inline src/PUReweight.cc]: Autumun --> Autumn

File: src/PUReweight.cc [hunk 11]
```diff
@@ -101,6 +113,123 @@ float PUReweight::weight(int MC, int target, int input, TString filename) {
...
+  npuAutumn18_59c7fb[93] = 0;	       
+  npuAutumn18_59c7fb[94] = 9.07281e-09;
+  npuAutumn18_59c7fb[95] = 0;	       
+  npuAutumn18_59c7fb[96] = 0;	       
+  npuAutumn18_59c7fb[97] = 0;	       
+  npuAutumn18_59c7fb[98] = 0;	       
+  npuAutumn18_59c7fb[99] = 0;		
+  
+  for(int k = 0 ; k < 100 ; ++k)
+  	hT2018_MCAutumun18_Data59c7fb.push_back(npuAutumn18_59c7fb[k]) ;
```
Comment (francescobrivio) [inline src/PUReweight.cc]: Autumun --> Autumn

File: src/PUReweight.cc [hunk 12]
```diff
@@ -101,6 +113,123 @@ float PUReweight::weight(int MC, int target, int input, TString filename) {
...
+
+  // RUN2ANALYSIS MC Autumn2018 target Data 59.7/fb - Full Run 2018
+  // Special weights for each sample read from external txt file
+  if (filename == TString(""))
+  {
+    for(int k = 0 ; k < 100 ; ++k)
+        hT2018_MCAutumun18_Data59c7fb_special.push_back(0.0) ; // if no filename passe, initialize a vector of 0s
```
Comment (francescobrivio) [inline src/PUReweight.cc]: Autumun --> Autumn

File: src/PUReweight.cc [hunk 13]
```diff
@@ -101,6 +113,123 @@ float PUReweight::weight(int MC, int target, int input, TString filename) {
...
+  }
+  else
+    hT2018_MCAutumun18_Data59c7fb_special = LoadExternalWeights(theFilename);
```
Comment (francescobrivio) [inline src/PUReweight.cc]: Autumun --> Autumn

File: src/TauTriggerSFs2017.cc [hunk 14]
```diff
@@ -70,7 +78,7 @@ TauTriggerSFs2017::TauTriggerSFs2017(const std::string& inputFileName, const std
   fitMCMap_ [ 0] = loadTF1(inputFile_, Form("%s_%s%s_dm0_MC_fit", trigger_.data(), tauWP_.data(), wpType_.data()));
   fitMCMap_ [ 1] = loadTF1(inputFile_, Form("%s_%s%s_dm1_MC_fit", trigger_.data(), tauWP_.data(), wpType_.data()));
   fitMCMap_ [10] = loadTF1(inputFile_, Form("%s_%s%s_dm10_MC_fit", trigger_.data(), tauWP_.data(), wpType_.data()));
-
+
```
Comment (francescobrivio) [inline src/TauTriggerSFs2017.cc:81]: delete the two spaces

File: test/skimNtuple.cpp [hunk 15]
```diff
@@ -1017,8 +1034,11 @@ int main (int argc, char** argv)
   for (int i = 0 ; i < 2; i++)
         myIDandISOScaleFactor[i] = new ScaleFactor();
 
-  myIDandISOScaleFactor[0] -> init_ScaleFactor("weights/HTT_SF_2016/Muon/Run2017/Muon_IdIso_IsoLt0.15_eff_RerecoFall17.root");
-  myIDandISOScaleFactor[1] -> init_ScaleFactor("weights/HTT_SF_2016/Electron/Run2017/Electron_IdIso_IsoLt0.10_eff_RerecoFall17.root");
+  //myIDandISOScaleFactor[0] -> init_ScaleFactor("weights/HTT_SF_2016/Muon/Run2017/Muon_IdIso_IsoLt0.15_eff_RerecoFall17.root");
+  //myIDandISOScaleFactor[1] -> init_ScaleFactor("weights/HTT_SF_2016/Electron/Run2017/Electron_IdIso_IsoLt0.10_eff_RerecoFall17.root");
+
+  myIDandISOScaleFactor[0] -> init_ScaleFactor("weights/idIso_SF_2018/Muon_Run2018_IdIso.root");
```
Comment (francescobrivio) [inline test/skimNtuple.cpp]: move this weight file to "/weights/HTT_SF_2016/Muon/Run2018/" there is no need to create a new "idIso_SF_2018" folder

File: test/skimNtuple.cpp [hunk 16]
```diff
@@ -1017,8 +1034,11 @@ int main (int argc, char** argv)
...
+  myIDandISOScaleFactor[1] -> init_ScaleFactor("weights/idIso_SF_2018/Electron_Run2018_IdIso.root");
```
Comment (francescobrivio) [inline test/skimNtuple.cpp]: move this weight file to "/weights/HTT_SF_2016/Electron/Run2018/" there is no need to create a new "idIso_SF_2018" folder

File: test/skimNtuple.cpp [hunk 17]
```diff
@@ -1640,6 +1724,11 @@ int main (int argc, char** argv)
       int t1hs = -1;
       int t2hs = -1;
 
+      int idx1hs_b = -1;     // bjet-1 index     // FRA DEBUG  ##QUI##
+      int idx2hs_b = -1;     // bjet-2 index                   ##QUI##
+      TLorentzVector vGenB1; // bjet-1 tlv                     ##QUI##
+      TLorentzVector vGenB2; // bjet-2 tlv                     ##QUI##
```
Comment (francescobrivio) [inline test/skimNtuple.cpp]: remove all the "##QUI##"

File: test/skimNtuple.cpp [hunk 18]
```diff
@@ -1658,6 +1747,7 @@ int main (int argc, char** argv)
 	      bool isFirst     = CheckBit (theBigTree.genpart_flags->at(igen), 12) ; // 12 = isFirstCopy
 	      bool isLast      = CheckBit (theBigTree.genpart_flags->at(igen), 13) ; // 13 = isLastCopy
 	      bool isHardScatt = CheckBit (theBigTree.genpart_flags->at(igen), 5) ; //   3 = isPromptTauDecayProduct
+	      bool isHardProcess = CheckBit (theBigTree.genpart_flags->at(igen), 7) ; //  7 = isHardProcess, for b coming from H  ##QUI##
```
Comment (francescobrivio) [inline test/skimNtuple.cpp]: remove "##QUI##"

File: test/skimNtuple.cpp [hunk 19]
```diff
@@ -1727,7 +1817,20 @@ int main (int argc, char** argv)
 		      // cout << "THIS: " << pdg << " px=" << theBigTree.genpart_px->at(igen) << endl;
 		    }
 		}
+		
+	       // FRA DEBUG - find the bjets from the Higgs decay     ##QUI##
```
Comment (francescobrivio) [inline test/skimNtuple.cpp]: remove "##QUI##"

File: test/skimNtuple.cpp [hunk 20]
```diff
@@ -1785,6 +1888,16 @@ int main (int argc, char** argv)
 	  mHH = vSum.M();
 	  vH1.Boost(-vSum.BoostVector());                     
 	  ct1 = vH1.CosTheta();
+	  
+	        // FRA DEBUG - build gen b jets       ##QUI##
```
Comment (francescobrivio) [inline test/skimNtuple.cpp]: remove "##QUI##"

File: test/skimNtuple.cpp [hunk 21]
```diff
@@ -3488,6 +3612,15 @@ int main (int argc, char** argv)
         // Now that I've selected the bjets build the TLorentzVectors
         TLorentzVector tlv_firstBjet (theBigTree.jets_px->at(bjet1idx), theBigTree.jets_py->at(bjet1idx), theBigTree.jets_pz->at(bjet1idx), theBigTree.jets_e->at(bjet1idx));
         TLorentzVector tlv_secondBjet(theBigTree.jets_px->at(bjet2idx), theBigTree.jets_py->at(bjet2idx), theBigTree.jets_pz->at(bjet2idx), theBigTree.jets_e->at(bjet2idx));
+	
+	// ##QUI## --> QUI PUOI VEDERE UN ESEMPIO DI COME CALCOLARE LE EFFICIENZE CHE IN QUESTO CASO SONO SALVATE SU DEI BRANCH NUOVI CHE AVEVO AGGIUNTO
```
Comment (francescobrivio) [inline test/skimNtuple.cpp]: remove this comment line

**Review by francescobrivio — CHANGES_REQUESTED** [review]

**Comment by dzuolo** [comment]
All requested changes implemented.

**Pull request merged**